asim: validate to real cluster rebalancing

kvoli opened the pull request:
This patch resolves some issues in the allocator that relate to determinism and accounting.

Various fixes for simulation determinism and correctness are then addressed.

resolves #83989
resolves #83990
resolves #89394
resolves #89397

kvoli committed:
kvserver: fix allocator determinism

Previously, the allocator when running in deterministic mode, would not
be deterministic under all scenarios due to iteration over an unordered
map. This patch fixes this by first sorting entries, before iterating.

resolves: cockroachdb#89394

Release note: None

kvoli committed:
kvserver: update storepool after relocate range

Previously, after a range relocation via the store rebalancer, no update
to the storepool would be made. It was therefore possible to have
conflicting views of the cluster capacity between the allocator, picking
targets and the store rebalancer loop.

This patch adds updates to the storepool on a store rebalancer relocate
range. This also updates the store rebalancer to "refresh" the rebalance
context with the latest storepool values, rather than maintain custom
logic locally.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/89771)
<!-- Reviewable:end -->

kvoli committed:
split: add method to determine if finder is ready

This patch adds a new method to the load based split decider:
`IsFinderReady`. This method may be used to determine if the finder
associated with the decider is ready to return a split key.

Release note: None

kvoli committed:
replicastats: adhere to min stats duration

Previously, the min stats duration was not observed when returning a
rated summary from replica stats. This patch modifies the creation of
the rated summary to return 0 when the record duration is less than the
min stats duration (5s).

Release note: None

kvoli committed:
asim: make simulator deterministic and valid

Previously, it was possible for the simulator to output simulated runs
that were inconsistent with a real cluster, w.r.t rebalancing.

This patch addresses this by updating several components and adding
additional observability.

The simulation was also non-deterministic, this is fixed in this patch.

Release note: None

kvoli commented:
I've trimmed down this patch quite a bit. Removing extra stuff that could be added later or didn't immediately seem necessary.

It now just consists of various fixes in the allocator and simulator that make the simulated runs both valid and deterministic.

I removed the roachtest for now as it was bloating the PR.

kvoli commented on pkg/kv/kvserver/store_rebalancer.go:
```diff
@@ -425,6 +424,8 @@ func (sr *StoreRebalancer) RebalanceLeases(
 		return NoRebalanceNeeded, candidateReplica, target
 	}
 
+	//log.Infof(ctx, "storemap state %v\n storepool state %v\n", rctx.)
```
remove

kvoli commented on pkg/kv/kvserver/store_rebalancer.go:
```diff
@@ -458,22 +459,25 @@ func (sr *StoreRebalancer) applyLeaseRebalance(
+// updated.
+func (sr *StoreRebalancer) RefreshContext(rctx *RebalanceContext) {
+	if localDesc, ok := sr.allocator.StorePool.GetStoreDescriptor(rctx.LocalDesc.StoreID); ok {
+		rctx.LocalDesc = &localDesc
```
Grab out of all stores list to get consistent view.

kvoli commented on pkg/kv/kvserver/split/decider.go:
```diff
@@ -186,6 +186,14 @@ func (d *Decider) recordLocked(
 
+// IsFinderReady returns true if the decider has an initialized finder, having
+// exceeded the split threshold recently and the finder is ready to find a key.
+func (d *Decider) IsFinderReady(now time.Time) bool {
```
Drop this commit.

kvoli commented:
superseded
